x/concentrated-liquidity/fees_test.go

```diff
@@ -448,3 +449,229 @@ func (suite *KeeperTestSuite) TestChargeFee() {
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     ownerWithValidPosition,
+			lowerTick: 0,
+			upperTick: 1,
+
+			currentTick: -1,
+
+			expectedFeesClaimed: sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(10))),
+		},
+
+		// error cases.
+
+		"invalid pool id given": {
+			initialLiquidity: sdk.OneDec(),
+
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     ownerWithValidPosition,
+			lowerTick: 0,
+			upperTick: 1,
+
+			currentTick: 2,
+
+			isInvalidPoolIdGiven: true,
+			expectedError:        cltypes.PoolNotFoundError{PoolId: 2},
+		},
+		"position does not exist": {
+			initialLiquidity: sdk.OneDec(),
+
+			lowerTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(0))),
+			upperTickFeeGrowthOutside: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			globalFeeGrowth: sdk.NewDecCoins(sdk.NewDecCoin(ETH, sdk.NewInt(10))),
+
+			owner:     s.TestAccs[1], // different owner from the one who initialized the position.
+			lowerTick: 0,
+			upperTick: 1,
+
+			currentTick: 2,
+
+			expectedError: cltypes.PositionNotFoundError{PoolId: 1, LowerTick: 0, UpperTick: 1},
+		},
+	}
+
+	for name, tc := range tests {
+		s.Run(name, func() {
+			tc := tc
+			s.SetupTest()
+
+			validPool := s.PrepareConcentratedPool()
+			validPoolId := validPool.GetId()
+
+			s.FundAcc(validPool.GetAddress(), tc.expectedFeesClaimed)
+
+			clKeeper := s.App.ConcentratedLiquidityKeeper
+			ctx := s.Ctx
+
+			err := clKeeper.InitializeFeeAccumulatorPosition(ctx, validPoolId, ownerWithValidPosition, tc.lowerTick, tc.upperTick, tc.initialLiquidity)
+			s.Require().NoError(err)
+
+			s.initializeTick(ctx, tc.lowerTick, tc.initialLiquidity, tc.lowerTickFeeGrowthOutside, false)
+
+			s.initializeTick(ctx, tc.upperTick, tc.initialLiquidity, tc.upperTickFeeGrowthOutside, true)
+
+			validPool.SetCurrentTick(sdk.NewInt(tc.currentTick))
+			clKeeper.SetPool(ctx, validPool)
+
+			err = clKeeper.ChargeFee(ctx, validPoolId, tc.globalFeeGrowth[0])
+			s.Require().NoError(err)
+
+			poolBalanceBeforeCollect := s.App.BankKeeper.GetBalance(ctx, validPool.GetAddress(), ETH)
+			ownerBalancerBeforeCollect := s.App.BankKeeper.GetBalance(ctx, tc.owner, ETH)
+
+			sutPoolId := validPoolId
+			if tc.isInvalidPoolIdGiven {
+				sutPoolId = sutPoolId + 1
+			}
+
+			// System under test
+			actualFeesClaimed, err := clKeeper.CollectFees(ctx, sutPoolId, tc.owner, tc.lowerTick, tc.upperTick)
+
+			// Assertions.
+
+			if tc.expectedError != nil {
+				s.Require().Error(err)
+				s.Require().ErrorAs(err, &tc.expectedError)
```

Comment:
can we also add assertions and checks for checking that position remains unchanged when erroring

Reply:
+ check for pool & owner balance

Reply:
Done: [52e97f9](https://github.com/osmosis-labs/osmosis/pull/3990/commits/52e97f98bb78e6652e0621bbdaa981811e5bfb7c)